Bug 1557293: audit log plugin truncates sqlquery to 512 bytes

gl-sergei opened the pull request:
Instead of using fixed-length buffer for escaped query, escape all
strings and do other formatting into single buffer of size 4096
bytes. When 4096 bytes is not enough, allocate larger buffer (one per
thread).

gl-sergei commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1156/#showFailuresLink

Only difference between this PR and tested branch is that commits are squashed in PR.

gl-sergei commented on plugin/audit_log/audit_log.c:
```diff
@@ -411,40 +450,48 @@ size_t audit_log_general_record(char *buf, size_t buflen,
+                    MAX_RECORD_ID_SIZE - MAX_TIMESTAMP_SIZE)
+  {
+    return full_outlen * 2 + strlen(format_string[audit_log_format]) +
+           MAX_RECORD_ID_SIZE - MAX_TIMESTAMP_SIZE;
```
Both condition and return statement need some adjustments. Here is better revision:

``` c++
  size_t buflen_estimated= full_outlen * 2 +
                           strlen(format_string[audit_log_format]) +
                           strlen(name) +
                           strlen(event->general_sql_command.str) +
                           20 + /* general_thread_id */
                           20 + /* status */
                           MAX_RECORD_ID_SIZE + MAX_TIMESTAMP_SIZE;
  if (buflen_estimated > buflen)
  {
    return buflen_estimated;
  }
```

gl-sergei commented:
See also
https://bugs.launchpad.net/percona-server/+bug/1588439
https://bugs.launchpad.net/percona-server/+bug/1588447

gl-sergei commented:
Addressed my own comments
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1165/

laurynas-biveinis commented on mysql-test/include/audit_log_events.inc:
```diff
@@ -28,7 +28,7 @@ connection default;
 create table t2 (id int);
 insert into t1 values (1), (2);
-insert into t2 values (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2 ...
+insert into t2 values (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2), (1), (2 ...
```
Is this query actually longer than 4K? (hard to tell looking at it - maybe add a comment saying so?)

gl-sergei replied:
Yes. It is :) I also added test case to log multiple larger queries.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -242,15 +242,37 @@ void csv_escape(const char *in, size_t *inlen, char *out, size_t *outlen)
+  @param[out] endptr       A pointer to the character after the
+                           last escaped character in the output
+                           buffer
+  @param[out] full_outlen  Full lenght of the output buffer needed to
```
s/lenght/length

laurynas-biveinis replied:
I'd say something like "Length of the output buffer that would be needed to store complete non-truncated escaped input buffer"

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -264,6 +286,22 @@ char *escape_string(const char *in, size_t inlen,
+  if (full_outlen)
+  {
+    char tmp[20];
+    in= in + inlen;
```
Input combination in == NULL, ful_outlen != NULL would crash?

laurynas-biveinis replied:
Could use +=, -= here and in the next few lines a lot

gl-sergei replied:
`in == NULL, ful_outlen != NULL` currently can potentially occur for query events. Not sure why it didn't...

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -264,6 +286,22 @@ char *escape_string(const char *in, size_t inlen,
+      format_escape_func[audit_log_format](in, &inlen, tmp, &tmp_size);
+      in= in + inlen;
+      inlen_orig= inlen_orig - inlen;
+      *full_outlen= *full_outlen + outlen;
```
Shoudn't this be "*full_outlen += tmp_size" ?

gl-sergei replied:
Good catch! Actually adding outlen could even make harm in case of short strings!

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -411,40 +450,51 @@ size_t audit_log_general_record(char *buf, size_t buflen,
+                   make_timestamp(timestamp, sizeof(timestamp), t),
+                   event->general_sql_command.str,
+                   event->general_thread_id,
+                   status, query, user, host, external_user,ip);
```
, ip

laurynas-biveinis replied:
Can we end up in a situation where buffer was enough to escape all the variables above, but not enough to snprintf without truncation?

gl-sergei replied:
This
`if (buflen_estimated > buflen)
    return buflen_estimated;`
prevents us from this situation.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -493,34 +543,34 @@ size_t audit_log_connection_record(char *buf, size_t buflen,
+  database= escape_string(event->database, event->database_length,
+                          endptr, endbuf - endptr, &endptr, NULL);
+
+  len= my_snprintf(endptr, endbuf - endptr,
```
If the buffer is guaranteed to be big enough, can you add an assert for this?

gl-sergei replied:
Yes. It is should be enough. USER, IP, HOST, DATABASE all have strictly limited length.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -358,14 +395,16 @@ size_t audit_log_audit_record(char *buf, size_t buflen,
 
 static
-size_t audit_log_general_record(char *buf, size_t buflen,
+size_t audit_log_general_record(char **buf, size_t buflen,
```
"buf" could use a header comment

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -697,12 +769,21 @@ void audit_log_notify(MYSQL_THD thd __attribute__((unused)),
                                     event_general->general_error_code,
                                     event_general);
-      audit_log_write(buf, len);
+      if (len > sizeof(buf))
```
Is get_large_buffer() call expensive in its non-allocating path? If cheap, then for every connection that has a buffer allocated already, I'd just use that buffer instead of the stack one on the first attempt

gl-sergei replied:
Should not be expensive. Basically it `(uchar*)thd->variables.dynamic_variables_ptr + offset` with additional checks.

laurynas-biveinis commented on plugin/audit_log/buffer.c:
```diff
@@ -163,11 +167,25 @@ void audit_log_buffer_shutdown(audit_log_buffer_t *log)
+  {
+    if (!log->drop_if_full)
+    {
+      mysql_mutex_lock(&log->mutex);
```
If one runs with big queries and too small log buffer, they will be in this mode constantly even though configuration will say "async", with performance surprises. Or maybe user could bump the log buffer a bit and it will fit its queries, but he will never learn this. Does it make sense to diagnose this somehow? Not in the scope of this PR

gl-sergei replied:
We can write message to the error log. Something like "audit_log_buffer size N is too small. Please consider increasing it to at least M.". We can write it every time when M breaks new record or once in a period or both.

laurynas-biveinis replied:
What user would need is means to ascertain whether the problem needs addressing or is one-of. I'd probably add a status counter and diagnose the first occurrence to the error log. Obviously not in this PR

gl-sergei commented:
changes:
- added test case for queries of various length
- changed comment for full_length
- correct handling of case when full_length != && in == NULL (which is theoretically possible)
- changed audit_log_*_record interface
- always use allocated buffer if available
- rename large_buffer to record_buffer (large_buffer will mess with audit_log_buffer)
- assert in audit_log_connection_record that buffer has enough space
- fixed `*full_outlen += tmp_size` bug

gl-sergei commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1179/

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -411,40 +457,54 @@ size_t audit_log_general_record(char *buf, size_t buflen,
+                   make_timestamp(timestamp, sizeof(timestamp), t),
+                   event->general_sql_command.str,
+                   event->general_thread_id,
+                   status, query, user, host, external_user, ip);
```
assert *outlen == buflen_estimated ?

gl-sergei replied:
Well, actually `*outlen == buflen_estimated` is a wrong assert.
This one should work:

``` c++
  *outlen= snprintf(endptr, endbuf - endptr,
                    format_string[audit_log_format],
                    name,
                    make_record_id(id_str, sizeof(id_str)),
                    make_timestamp(timestamp, sizeof(timestamp), t),
                    event->general_sql_command.str,
                    event->general_thread_id,
                    status, query, user, host, external_user, ip);

  DBUG_ASSERT(endptr + *outlen <= buf + buflen);
```

I replaced `my_snprintf` with `snprintf` because it seems that `my_snprintf` returns the number of characters printed, and `snprintf` returns the number of characters that would have been printed if the size was unlimited.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -909,9 +1032,52 @@ static struct st_mysql_sys_var* audit_log_system_variables[] =
+  {
+    local= (audit_log_thd_local *)
+            my_malloc(sizeof(audit_log_thd_local), MYF(MY_FAE));
+    memset(local, 0, sizeof(audit_log_thd_local));
```
my_malloc + memset -> my_malloc (MYF(... | MY_ZEROFILL));

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -909,9 +1032,52 @@ static struct st_mysql_sys_var* audit_log_system_variables[] =
+    if (buf == NULL)
+      buf= (char *) my_malloc(size, MYF(MY_FAE));
+    else
+      buf= (char *) my_realloc(buf, size, MYF(MY_FAE));
```
Can replace if with buf= my_realloc( ... MYF(MY_FAE | MY_ALLOW_ZERO_PTR));

gl-sergei commented:
changes:
- added assert to audit_log_*_record functions
- changed my_realloc and my_malloc as suggested

http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1182/

laurynas-biveinis commented:
LGTM, please log some bug to track diagnosing the sync logging in async mode condition

gl-sergei commented:
Here is a bug report: https://bugs.launchpad.net/percona-server/+bug/1590907
